Update Gruneisen EOS

jhp-lanl opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "fix bug in ideal gas EOS.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary

Numerous changes to the Gruneisen EOS

### Changes:

- Redoes how the Gruneisen EOS calculates the pressure and bulk modulus to allow specification of $P_0$. I also cast the equations in something more understandable I think.
- Updated documentation with to better mirror the code and provide a better description of $\Gamma$ vs $\gamma$ in this context
- Added a `rho_max` variable to the model that allows the user to specify a maximum compression for the EOS. Alternatively, if this is not specified, the model will calculate the density limit based off of the singularity that exists in the denominator of the pressure expression. Lookups beyond this density will just return the values at the limit, which should be sufficiently large to reasonably approximate the compressibility limit. Finding the limit for the linear Hugoniot case is trivial, but the cubic case required significantly more effort.
- Split up the `eos_unit_tests.cpp` file into multiple files for organization and to speed up compilation times when the Catch2 hooks have already been compiled.
- Add new Gruneisen EOS tests to ensure that the EOS both reproduced the jump conditions as well as produces a bulk modulus that is consistent with a finite difference approximation.

NOTE: The temperature and the EOS in expansion aren't thermodynamically consistent. This will need to be addressed in the future.

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.

jhp-lanl commented:
@chadmeyer I'm sure you're sick of reviewing my EOS changes but I think you're still the best one to review here.

These changes were mainly necessary to make our implementation operable in Pagosa. Specifically, the P_0 parameter wasn't actually doing anything in our implementation but was used in Pagosa's. I've also decided to adopt the nomenclature of `eostype(m) = 11` in xRAGE since it's neater and I think demonstrates the Mie-Gruneisen form better.

There were diffs but they were in the last two decimals for pressure. The bulk modulus differences were larger, but were expected since the bulk modulus approximately scales with the square root of the pressure.

jhp-lanl commented:
> I just worry that without GPU test coverage, it'll be easy to make a change that causes a segfault or something down the line.

Fair enough. I may just loop it then rather than checking the full vector functionality.

Aiming for full code coverage is always my goal, so I should have just not been lazy :-)

Yurlungur commented:
Looping should be fine, since it inherits from the full vector stuff, which we already know works.

jhp-lanl commented:
@Yurlungur I made it run on GPUs and decided to just use the same strategy as the vector EOS test since it seemed the most similar to what I was already doing.

If there's an easier way to just copy one value at a time in a for loop, I'd be happy to change the code to do that. But at least it runs on the GPU now.

Yurlungur commented:
There's a way to do that @jhp-lanl using either deep copies or `parReduce` but this is fine. I don't think it's any cleaner than what you've done.

Yurlungur commented:
I would like to wait until @chadmeyer reviews before merging if possible.

chadmeyer commented:
It would seem that there is no ability in github to comment on unchanged lines, so I'm just going to quote the section:
```c++
PORTABLE_FUNCTION void Gruneisen::DensityEnergyFromPressureTemperature(
    const Real press, const Real temp, Real *lambda, Real &rho, Real &sie) const {
  sie = _Cv * (temp - _T0);
  // We have a branch at rho0, so we need to decide, based on our pressure, whether we
  // should be above or below rho0
  Real Pref = PressureFromDensityInternalEnergy(_rho0, sie);
  if (press < Pref) {
    rho = (press + _C0 * _C0 * _rho0) / (_C0 * _C0 + _G0 * sie); <=== THIS LINE IS PROBABLY WRONG NOW!
  } else { // We are in compression; iterate
    auto residual = [&](const Real r) {
      return press - PressureFromDensityInternalEnergy(r, sie);
    };
    Real rho1 = _rho0, res1 = residual(rho1), slope = _G0 * sie + _C0 * _C0, rho2, res2,
         rhom, resm;
    rho2 = (rho > rho1 + 1e-3) ? rho : rho1 + res1 / slope;
    res2 = residual(rho2);
    for (int i = 0; i < 20; ++i) {
      slope = (rho2 - rho1) / (res2 - res1 + 1.0e-10);
      rhom = rho2 - res2 * slope;
      resm = residual(rhom);
      if (resm / press < 1e-8) break;
      rho1 = rho2;
      res1 = res2;
      rho2 = rhom;
      res2 = resm;
    }
    rho = rhom;
  }
}
```

I think the second half of the routine is OK because it iterates using functions that contain pressure and the `_P0` parameter, but I think this particular branch will be wrong.  @jhp-lanl, do you agree?

jhp-lanl commented:
> I think the second half of the routine is OK because it iterates using functions that contain pressure and the _P0 parameter, but I think this particular branch will be wrong. @jhp-lanl, do you agree?

@chadmeyer Good catch! Let me see what I can do here to reduce code duplication and make this consistent with my changes. I'll add a test as well.

Yurlungur commented:
I assume we're waiting on this until we resolve current questions downstream, @jhp-lanl @annapiegra ?

jhp-lanl commented:
> I assume we're waiting on this until we resolve current questions downstream, @jhp-lanl @annapiegra ?

We're waiting on a definitive formulation for the bulk modulus since it seems like the current formulations fail to reproduce the finite difference solution. We're also waiting on me since I've been busy with other things, but I'm going to be working on this tomorrow.

jhp-lanl commented:
The last thing to resolve on this MR is the introduction of a maximum density for the EOS. I'll do this next week

jhp-lanl commented:
@jonahm-LANL and @chadmeyer I've introduced the maximum density limit in the EOS now. It was made a bit complicated by the fact that we can have up to a cubic fit for the Hugoniot which could lead to multiple compression limits. Some of these limits may be excluded by other properties of the EOS, but instead of trying to run down all those limits, I figured it was easier to just naively assume that any fit could be given and that it could produce a compression limit for the EOS. If you guys can review the algorithm, I'm fairly certain (due to all the tests I've added) that the compression limit code is doing what it's supposed to.

@annapiegra and @mauneyc-LANL this will add the capability to introduce a maximum density limit if you want to merge it into your testing. I also have also added a few tests that make me fairly certain the pressure and bulk modulus are accurate in singularity: I have one test comparing the calculated bulk modulus to what you would get with finite differences, and I have another test comparing a state on the reference curve to what you would get from the Rankine-Hugoniot jump conditions. Both tests pass, which makes me fairly confident in the implementation we have here.

Yurlungur commented on doc/sphinx/src/models.rst:
```diff
@@ -214,40 +230,66 @@ The user should note that this implies that :math:`e=0` at the reference
 
-and when the unitless user parameter :math:`b=0`, this ensures the the Gruneisen
+When the unitless user parameter :math:`b=0`, this ensures the the Gruneisen
 parameter is of a form where :math:`\rho\Gamma =` constant in compression.
```
Given that compression is now a variable, this phrasing has become confusing I think. Maybe "in the so-called compression limit, when when eta < 0"

Yurlungur replied:
I think including rho Gamma = constant is good, I just want to reference what "in compression" means here.

jhp-lanl replied:
Ah gotcha. That makes sense! I'll change that

Yurlungur commented on singularity-eos/eos/eos.hpp:
```diff
@@ -125,10 +125,19 @@ class Gruneisen : public EosBase<Gruneisen> {
             const Real b, const Real rho0, const Real T0, const Real P0, const Real Cv)
       : _C0(C0), _s1(s1), _s2(s2), _s3(s3), _G0(G0), _b(b), _rho0(rho0), _T0(T0), _P0(P0),
-        _Cv(Cv) {}
+        _Cv(Cv), _rho_max(0.99 * GetRhoMax(s1, s2, s3, rho0)) {}
```
I assume the 0.99 is a safety factor?

Yurlungur replied:
Might be good to define a static class variable `static constexpr Real RHOMAX_SAFETY = 0.99` or something... and maybe a comment indicating this is consistent with Pagosa.

Yurlungur commented on test/eos_unit_test_helpers.hpp:
```diff
@@ -0,0 +1,101 @@
+#ifndef _SINGULARITY_EOS_TEST_TEST_HELPERS_
+#define _SINGULARITY_EOS_TEST_TEST_HELPERS_
+
+#include "catch2/catch.hpp"
```
Missing `#ifndef CATCH_CONFIG_RUNNER` guard

Yurlungur commented on singularity-eos/eos/eos_gruneisen.cpp:
```diff
@@ -12,12 +12,155 @@
+  } else if (maxroot > min && maxroot < max) {
+    return maxroot;
+  } else {
+    return min - 1.;
```
how should I interpret this else case? Is this a fail case? If so perhaps we should return a specific flag number, like signal_nan?

jhp-lanl replied:
I might have coded this to be a little too general here, because my intent was to just fail at an arbitrary number outside of the bounds on the minimum side. Since I'm only operating on `eta` in compression, any negative value would suffice. I'll change this.

jhp-lanl replied:
> 👍 Maybe add a comment explaining the logic as well.

There is a comment at the top of the function explaining the behavior. Were you wanting something more verbose?

Yurlungur replied:
nah just a comment at this else case like "any negative number will do" or something like that.

Yurlungur commented on singularity-eos/eos/eos_gruneisen.cpp:
```diff
@@ -12,12 +12,155 @@
+      }
+      if (poly(minbound) * poly(maxbound) <= 0) {
+        // Root is appropriately bounded
+        using RootFinding1D::bisect;
```
I might try the new regula falsi method instead of bisection, thought it probably doesn't matter too much

jhp-lanl replied:
Yeah it wasn't available yet when I was coding this up. I just wanted to avoid secant really. I'll change this to the false position root finder

Yurlungur commented on singularity-eos/eos/eos_gruneisen.cpp:
```diff
@@ -12,12 +12,155 @@
+  // Polynomial from the denominator of the reference pressure curve:
+  auto poly = [=](Real eta) { return 1 - s1 * eta - s2 * square(eta) - s3 * cube(eta); };
+
+  // First find the eta root. A negative root indicates that there is no maximum density.
+  Real root = 0; // Non-sensical root means a root hasn't been found yet
+  Real discriminant, root1, root2;
+  if (s2 == 0 && s3 == 0 && s1 > 0) {
+    // Linear Us-up analytic root
+    root = 1 / s1;
+  } else if (s3 == 0 && s2 != 0) {
+    // Quadratic Us-up
+    discriminant = square(s1) + 4 * s2;
+    if (discriminant < 0) {
+      root = -1; // Imaginary roots, so no limit
+    } else {
+      root1 = (s1 + std::sqrt(square(s1) + 4 * s2)) / (-2 * s2);
+      root2 = (s1 - std::sqrt(square(s1) + 4 * s2)) / (-2 * s2);
+      root = find_min_bounded_val(root1, root2, 0, 1);
+    }
+  } else if (s3 != 0) {
+    // Cubic Us-up
+    discriminant = -18 * s3 * s2 * s1 - 4 * cube(-s2) + square(s2) * square(s1) -
+                   4 * s3 * cube(s1) - 27 * square(s3);
+    // Use discriminant to help find roots
+    if (discriminant == 0) {
+      // Easy analytical roots (probably not the case)
+      if (square(s2) == 3 * s3 * s1) {
+        // Single root with multiplicity of 3 (inflection point is root)
+        // Also guards against divide by zero
+        root = -s2 / 3 / s3;
+      } else {
+        // One root has multiplicity of 2 (local extremum is a root)
+        root1 = (-9 * s3 - s2 * s1) / 2 / (square(s2) - 3 * s3 * s1);
+        root2 = (-4 * s3 * s2 * s1 - 9 * square(s3) + cube(s2)) /
+                (-s3 * (square(s2) - 3 * s3 * s1));
+        root = find_min_bounded_val(root1, root2, 0, 1);
+      }
```
Given that this only gets called in initialization, is there a reason to do the analytic root finding at all? Maybe we could save some code and always use the numerical root finder?

If we did the numerical root finding always, we could use the analytic formulae as checks that everything is working properly in the unit tests.

jhp-lanl replied:
The reason I'm doing analytic root finding is because I need to guarantee that the  minimum root is found. I briefly looked into the numerics of this and decided that there was no guaranteed way to find the _minimum_ bounded root without more information.

Feel free to find a better way though.

jhp-lanl replied:
Are you talking about the cubic case specifically? I think the extrema equations are still valid when the discriminant is also zero so I could probably relax the condition to `if (discriminant >= 0) {` and eliminate that code. The good thing about writing that code is that I do at least test it so the tests should automatically tell me whether it's feasible to eliminate the code.

For the quadratic and linear cases I don't think you're saving much. For the quadratic case you'd still need to find the extremum anyway and then have a conditional for the sign of `s2` to determine whether you're greater than or less than the extremum.

Yurlungur replied:
Ah yeah I guess the extremum trick only works for cubic.

jhp-lanl replied:
I added some comments to the extremum trick to make it more apparent that the trick works since the roots must be in a region of the polynomial that is _decreasing_. This is because the polynomial is pinned to 1 at $\eta = 0$, so there must be a decreasing region between $\eta = 0$ and $\eta = 1$ for there to be a possibility of any roots.

I also realized another reason to delete the analytic root part: the roots will only have multiplicity (i.e. the determinant is zero) when an extremum _**is**_ a root anyway. So I'm now requiring an explicit sign change across the bounds and otherwise testing the bounds themselves since the iterative approaches have a hard time converging if the root is an extremum.

Yurlungur replied:
That sounds good! We should double check it all works in the tests, of course. We can use the code you just replaced as a sanity check. :)

jhp-lanl replied:
I believe I already have tests that capture the behavior of the previous code so in that sense I think it works.

Interestingly the tests for root multiplicity (i.e. roots are extrema) failed when using the iterative root find because the tolerances were too tight. The main issue is that the root find tolerance is too loose for this particular situation where the dy is basically zero. I would have assumed that regula falsi would converge directly to a bound if that bound is a root though, so I'm not quire sure what's going on.

Either way, skipping the root find when one of the bounds is a solution is probably the intelligent thing to do here anyway.

Yurlungur replied:
Yeah that seems fine. My experience with Regula Falsi has been it's a little sensitive compared to bisection or secant actually. It's very fast, but it seems to fail out more frequently.

Yurlungur commented on singularity-eos/eos/eos_gruneisen.cpp:
```diff
@@ -12,12 +12,155 @@
+The Gruneisen EOS diverges at a specific compression. Ensure that the maximum density is
+below the smallest singularity in the reference pressure curve
+*/
+PORTABLE_FUNCTION Real Gruneisen::GetRhoMax(const Real s1, const Real s2, const Real s3,
```
It occurs to me that querying what rhomax is from the eos might be desirable but we don't want to re-do thhis calculation. I suggest renaming this `ComputeRhoMax` but exposing a method that return the value stored in the class for debugging purposes.

Yurlungur commented on singularity-eos/eos/eos_gruneisen.cpp:
```diff
@@ -12,12 +12,155 @@
+      Real minbound = 0;
+      Real maxbound = 1;
+      if (discriminant > 0) {
+        // Three real roots. We need to use the extrema to ensure we have a proper bracket
+        // in which to search for the root.
+        const Real extremum1 =
+            (2 * s2 + std::sqrt(square(-2 * s2) - 4 * 3 * s3 * s1)) / (-2 * 3 * s3);
+        const Real extremum2 =
+            (2 * s2 - std::sqrt(square(-2 * s2) - 4 * 3 * s3 * s1)) / (-2 * 3 * s3);
+        const Real min_extremum = std::min(extremum1, extremum2);
+        const Real max_extremum = std::max(extremum1, extremum2);
+        if (s3 < 0) {
```
I like the extrema trick to find the root you want. Nice idea.

Yurlungur commented on singularity-eos/eos/eos_gruneisen.cpp:
```diff
@@ -12,12 +12,155 @@
+    return rho0 / (1 - root);
+  } else {
+    // No bounded root exists so there is no upper-limit on the compression
+    return 1.e99;
```
should you return infinity here? Or will that cause problems?

jhp-lanl replied:
I just realized I never implemented the cutoff in the actually lookups 🤦 .

My intent was just to min the input density against the maximum. It looks like `std::numerical_limits::infinity` needs a special `max()` function. Is there a way to avoid this?

Yurlungur replied:
I think `std::min(rho, std::numerical_limits::infinity());` does what you want? Am I missing something?

jhp-lanl replied:
Yeah I don't know what I was thinking... I'll implement it

Yurlungur commented on test/test_eos_gruneisen.cpp:
```diff
@@ -0,0 +1,546 @@
+     unstable shocks (i.e. Us - up < c) but we admit these cases for the moment since
+     a more comprehensive set of bounds on the EOS parameters is not currently available.
+     For a linear Us-up relationship, the point at which up > Us occurs also corresponds
+     to the singularity in the
```
incomplete sentence

Yurlungur commented on test/test_eos_gruneisen.cpp:
```diff
@@ -0,0 +1,546 @@
+  }
+}
+
+SCENARIO("Aluminum Gruneisen EOS Sound Speed and Pressure Comparison", "[GruneisenEOS]") {
```
Would it be worth checking that the answer is reasonable way in compression? Maybe you're doing that?

jhp-lanl replied:
Not yet! I should put in a simple check that the values beyond max compression simply return the value at max compression.

jhp-lanl replied:
Test added

Yurlungur reviewed (approved):
Overall this looks good and these changes seem reasonable, and well tested to me. A few comments below. Also @chadmeyer should double check the physics reasoning.

jhp-lanl commented on doc/sphinx/src/models.rst:
```diff
@@ -214,40 +230,66 @@ The user should note that this implies that :math:`e=0` at the reference
 treat the temperature from this EOS as only a rough estimate.
 
-The Grunesien parameter is given by
+First defining the compression as
```
I think it's grammatically correct, but I'm using the passive voice pretty heavily. I'll figure out a way to make it more active

jhp-lanl commented on singularity-eos/eos/eos_gruneisen.cpp:
```diff
@@ -127,11 +270,24 @@ PORTABLE_FUNCTION void Gruneisen::DensityEnergyFromPressureTemperature(
   }
 }
-PORTABLE_FUNCTION void Gruneisen::FillEos(Real &rho, Real &temp, Real &sie, Real &press,
+PORTABLE_FUNCTION void Gruneisen::FillEos(Real &rho_in, Real &temp, Real &sie, Real &press,
                                           Real &cv, Real &bmod,
                                           const unsigned long output,
                                           Real *lambda) const {
   // The following could be sped up with work!
+  const unsigned long input = ~output;
+  if (thermalqs::temperature & input && thermalqs::pressure & input) {
+    DensityEnergyFromPressureTemperature(press, temp, lambda, rho_in, sie);
+  }
+  else if (thermalqs::density & output || thermalqs::specific_internal_energy & output) {
+    // Error out on density or energy output because they're currently required as inputs
+    std::stringstream errorMessage;
+    errorMessage << "Gruneisen FillEos: Density and energy are currently required inputs "
+                 << "except when pressure and temperature are inputs"
+                 << std::endl;
+    EOS_ERROR(errorMessage.str().c_str());
+  }
+  const Real rho = std::min(rho_in, _rho_max);
```
@jonahm-LANL note that I needed to modify the `FillEos` function a bit to add the density check. You may want to look at this as well.

This is related to issue https://github.com/lanl/singularity-eos/issues/72

mauneyc-LANL commented on singularity-eos/eos/eos_gruneisen.cpp:
```diff
@@ -12,90 +12,233 @@
+  // First find the eta root. A negative root indicates that there is no maximum density.
+  Real root = 0; // Non-sensical root means a root hasn't been found yet
+  Real discriminant, root1, root2;
+  if (s2 == 0 && s3 == 0 && s1 > 0) {
+    // Linear Us-up analytic root
+    root = 1 / s1;
+  } else if (s3 == 0 && s2 != 0) {
+    // Quadratic Us-up
+    discriminant = square(s1) + 4 * s2;
+    if (discriminant < 0) {
+      root = -1; // Imaginary roots, so no limit
+    } else {
+      root1 = (s1 + std::sqrt(discriminant)) / (-2 * s2);
+      root2 = (s1 - std::sqrt(discriminant)) / (-2 * s2);
+      root = find_min_bounded_val(root1, root2, 0, 1);
+    }
```
I'm always a bit nervous about math w/ implicit integer conversions
```c++
root = 1 / s1;
```
and testing FP equality w/ a literal integer `0`.

```c++
if (s2 == 0 && s3 == 0 && s1 > 0) {
```

I can also see a lot of issues with division by strictly "non-zero" FP numbers that are only a few bits different than `0`.

Should there be a tolerance on these conditions? e.g. `if (std::abs(s2)<TOL /* ... */ )`.
(I may be missing something key here, tho)

Yurlungur replied:
THis is a good catch. I second @mauneyc-LANL 's suggestion. Might be good to divide by denom + a little bit too.

jhp-lanl replied:
> I'm always a bit nervous about math w/ implicit integer conversions
> ```c++
> root = 1 / s1;
> ```

Is switching to
```c++
root = 1. / s1;
```
enough to alleviate this issue or should we define some of
```c++
constexpr Real ONE = 1.;
constexpr Real TWO = 1.;
constexpr Real HALF = 1.;
...
```
in `constants.hpp`?

> Should there be a tolerance on these conditions? e.g. if (std::abs(s2)<TOL /* ... */ ).
Sure that's not a bad idea. If that denominator is close to zero, I suppose it's probably good enough to use the linear case anyway and I'll make those comparisons to zero a bit more loose.

Yurlungur replied:
I think `1. / s1` should resolve the implicit conversion question. Let's not do `constexpr Real ONE = 1`.

I would be more nervous about divide by zero issues and exact comparison of real numbers with `==`. It's easy for floating point math to never exactly equal zero, so the tolerance trick is much more likely to be what we actually want.

Yurlungur replied:
So for `root = 1. / s1` I would actually do:
```C++
root = 1. / (s1 + MY_SIGN(s1)*EPS);
```
where `EPS` is some small number, like machine epsilon, and `MY_SIGN` evaluates to 1 if s1 >= 0 and -1 otherwise.

jhp-lanl replied:
Yeah I went with explicit comparisons because it's not a derived quantity but rather user-supplied parameters. I'd hope a user wouldn't try to provide a quadratic fit parameter when it's 1e-109, but who knows 🙃 . I'll make the changes shortly

Yurlungur replied:
I think you probably already know all this. But this discussion of floating points reminded me of some nice online resources on working with floating point numbers :)

- [What every computer scientist should know about floating-point arithmetic](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html)
- [Five tips for floating point programming](https://www.codeproject.com/Articles/29637/Five-Tips-for-Floating-Point-Programming)
- [Avoiding overflow and underflow](https://www.codeproject.com/Articles/25294/Avoiding-Overflow-Underflow-and-Loss-of-Precision)

mauneyc-LANL replied:
> I think `1. / s1` should resolve the implicit conversion question. Let's not do `constexpr Real ONE = 1`.
> 
> I would be more nervous about divide by zero issues and exact comparison of real numbers with `==`. It's easy for floating point math to never exactly equal zero, so the tolerance trick is much more likely to be what we actually want.

Offering an opinion: I'm generally in favor of named constants over literal constants, as it avoids issues with having to sort out when/how implict conversions happen. I'm sympathetic to the idea that `ONE` ect., is a bit much.

It may be worthwhile to have and use some number constants (a bit beyond the scope of this PR)

```c++
inline constexpr Real e          = /*...*/;
inline constexpr Real pi         = /*...*/;
inline constexpr Real ln2        = /*...*/;
inline constexpr Real sqrt2      = /*...*/;
inline constexpr Real inv_sqrt2  =/*...*/;
// and so on...
```

mauneyc-LANL replied:
> Is switching to
> 
> ```c++
> root = 1. / s1;
> ```
> 
> enough to alleviate this issue or should we define some of
> 
> ```c++
> constexpr Real ONE = 1.;
> constexpr Real TWO = 2.;
> constexpr Real HALF = ONE / TWO;
> ...
> ```
> 
> in `constants.hpp`?

using `1.` is usually what I default to doing

Yurlungur replied:
> > I think `1. / s1` should resolve the implicit conversion question. Let's not do `constexpr Real ONE = 1`.
> > I would be more nervous about divide by zero issues and exact comparison of real numbers with `==`. It's easy for floating point math to never exactly equal zero, so the tolerance trick is much more likely to be what we actually want.
> 
> Offering an opinion: I'm generally in favor of named constants over literal constants, as it avoids issues with having to sort out when/how implict conversions happen. I'm sympathetic to the idea that `ONE` ect., is a bit much.
> 
> It may be worthwhile to have and use some number constants (a bit beyond the scope of this PR)
> 
> ```c++
> inline constexpr Real e          = /*...*/;
> inline constexpr Real pi         = /*...*/;
> inline constexpr Real ln2        = /*...*/;
> inline constexpr Real sqrt2      = /*...*/;
> inline constexpr Real inv_sqrt2  =/*...*/;
> // and so on...
> ```

We actually have *some* of this, thanks to @brryan in the `singularity-eos/base/constants.hpp` file. It could be extended to include some particularly common non-rational numbers that aren't physical I would think.

Yurlungur replied:
Obviously that's a story for another time of course

mauneyc-LANL commented on singularity-eos/eos/eos_gruneisen.cpp:
```diff
@@ -12,90 +12,233 @@
+      //       necessarily be positive (the reverse does not hold). Thus the extrema below
+      //       are not imaginary.
+      const Real extremum1 =
+          (2 * s2 + std::sqrt(square(-2 * s2) - 4 * 3 * s3 * s1)) / (-2 * 3 * s3);
```
This is mostly for discussion, and am not really advocating for any change.

There could be some FLOPs saved here with a bit of simplification (may be wise to double-check me here)

```c++
ex1 = (s2 + std::sqrt(square(s2) - 3. * s3 * s1)/(-3. * s3);
```

and, if you wanna go REALLY crazy with it

```c++
auto threeS3 = 3. * s3;
auto I_threeS3 = 1. / threeS3; // division is expensive, so if there's reuse my instinct is to move to multipication
auto t1 = std::sqrt(square(s2) - threeS3 * s1);
Real ex1 = (s2 + t1) * I_threeS3;
Real ex2 = (s2 - t1) * I_threeS3;
```
This maybe a bit much, and likely as not most compilers are good enough in optimization here. There's also an argument to be made that in some contexts it's better to "load" FLOPS. On the other hand, keeping the FP arithmetic less "expressive" can help contain round-off and cancellation errors.

(side note: the quad formula is notorious for reaking havoc on precision, see Theorem 2 of https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html, https://www.johndcook.com/blog/2018/04/28/quadratic-formula/ for a nice summary).

Yurlungur replied:
good catch

mauneyc-LANL replied:
> (side note: the quad formula is notorious for reaking havoc on precision, see Theorem 2 of https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html, https://www.johndcook.com/blog/2018/04/28/quadratic-formula/ for a nice summary).

I see @Yurlungur got the jump on me with this

jhp-lanl replied:
This code should be just executed at initialization so I'm not too concerned about performance here, BUT I think your first suggestion enhances readability since it's a pretty small simplification that doesn't make it any less obvious what is being calculated here.

I think the second suggestion makes the code slightly less readable though. The fact that you were able to instantly recognize the quadratic formula I think is a good thing here.

jhp-lanl replied:
FYI - I think I can get away with losing precision in the discriminants of the quadratic formulas here for two reasons:
1.  The value of the extrema calculations don't need to be really precise since they are used to bracket the real root more appropriately.
2.  When I'm directly calculating the roots for the quadratic case, I can reasonably rely on `s1` to be of order unity. That way although I may lose precision in the discriminant if it's near zero, I don't really care since I'll basically get two very close estimates of the root. If the lower estimate is too conservative or too affected by machine precision issues, then we can re-examine this issue later.

Yurlungur commented on singularity-eos/eos/eos_gruneisen.cpp:
```diff
@@ -49,45 +52,48 @@ PORTABLE_FUNCTION Real Gruneisen::ComputeRhoMax(const Real s1, const Real s2, co
+  Real root = -1.; // Non-sensical root means none exists
   Real discriminant, root1, root2;
-  if (s2 == 0 && s3 == 0 && s1 > 0) {
+  if (is_near_zero(s2, EPS) && is_near_zero(s3, EPS) && s1 > EPS) {
```
are the s's all guaranteed to be positive? i.e., should this be `std::abs(s1) > EPS`?

jhp-lanl replied:
I would first argue that there is no guarantee since we don't enforce anything about the selection of the s values so I wouldn't assume anything about sign.

But in this case the critical test is that `s1` is positive, not that it is nonzero. If `s1` is negative, then the root would be negative anyway and it's not relevant.

However, you do remind me that if `s1 < 1`, the root would be greater than 1 anyway which is nonsensical. I'm not guarding against that though so really the condition should be `s1 > 1.`. I'll fix that

Yurlungur replied:
sounds good.

jhp-lanl commented:
I think I've addressed everybody's comments at this point and all tests pass both here and on gitlab. Let me know by tomorrow if you have other concerns.

chadmeyer commented on singularity-eos/eos/eos.hpp:
```diff
@@ -125,10 +125,19 @@ class Gruneisen : public EosBase<Gruneisen> {
  public:
   Gruneisen() = default;
   PORTABLE_INLINE_FUNCTION
+  Gruneisen(const Real C0, const Real s1, const Real s2, const Real s3, const Real G0,
+            const Real b, const Real rho0, const Real T0, const Real P0, const Real Cv,
+            const Real rho_max)
+      : _C0(C0), _s1(s1), _s2(s2), _s3(s3), _G0(G0), _b(b), _rho0(rho0), _T0(T0), _P0(P0),
+        _Cv(Cv), _rho_max(rho_max) {}
```
Should there be a check here in the constructor to ensure that user-specified rho_max is sane (and, for instance, less than the result of `ComputeRhoMax`)?

Yurlungur replied:
I would prefer to warn rather than die if rho_max is specified to be greather than `computeRhoMax`  but a warning would be good maybe.

jhp-lanl replied:
I'll create a hotfix for this shortly.

Just to be sure, @Yurlungur , we don't have a full-fledged warning mechanism other than print statements, right?

Yurlungur replied:
We don't. Here's my suggested design for now:
```C++
#ifndef NDEBUG
if (rho max condition violated) {
printf("some warning message);
}
#endif
```

I'll try to submit some more formal warning machinery as a PR at some point

jhp-lanl replied:
Updated here:
https://github.com/lanl/singularity-eos/pull/130
